### DISCUSSION TITLE
Arbitrary URL schemes for supported protocols?

### DISCUSSION THREAD

**Discussion by matt17r** [curl/curl#11700] [Q&A] [answered]
Is there a way to pass a URL containing an unsupported (or conflicting) scheme but tell curl to use a specific, supported protocol?

In my (admittedly frivolous) case I'd like to send `coffee://` requests to a HTCPCP server using the HTTP protocol. I can't come up with many (any?) other use cases, the only thing that comes to mind is perhaps wanting to send a HTTP request to an `ftp://` URL (which feels like a stretch).

I've read https://curl.se/dev/new-protocol.html and I don't think an April Fools joke (even one with an RFC) justifies a new protocol being added but I'm hoping there's some way to force the protocol or tell curl to ignore the scheme.

I tried forcing the protocol with `--proto =http` but I still get `Protocol "coffee" not supported or disabled in libcurl` (as expected once I read the docs).

Here's the command I'm trying to make work:

```
curl 'coffee://localhost:3000' \
  --proto =http \
  --request 'BREW' \
  --header  'Content-Type: message/coffeepot' \
  --header  'Accept-Additions: milk-type/Cream, syrup-type/Vanilla' \
  --data    'coffee-message-body=start'
```

**Comment by bagder** [comment, marked as answer]
No, the URL scheme is extracted and checked for and if not one of the built-in supported ones, curl will simply report an error. There is no logic in curl that allows the scheme to get replaced somehow, like it can do for the host name for example. I can't think of a good reason to provide such a replacement/pattern matching system for the URL scheme.

You could simply skip the scheme and `localhost:3000#coffee` or something, as then curl will "guess" which scheme to use an default to HTTP and then the rest of the command line should work.

**Reply by matt17r** [reply]
Thanks for confirming there isn't a way around it. For my purposes (implementing a demonstration HTCPCP server) I've removed the scheme from the URL and added it as a custom header:

```
> curl 'localhost:3000' \
  --request 'BREW' \
  --header  'Scheme: coffee' \
  --header  'Content-Type: message/coffeepot' \
  --header  'Accept-Additions: milk-type/Cream, syrup-type/Vanilla' \
  --data    'coffee-message-body=start'
```

Then in my Rack app, if `env["HTTP_SCHEME"]` is set I replace Rack's `url_scheme` with the contents of the `HTTP_SCHEME` header.
